aten/src/ATen/BatchedFallback.cpp

```diff
@@ -34,6 +34,172 @@ static bool areAnyArgumentsTensorList(const FunctionSchema& schema) {
       [] (const Argument& arg) { return arg.type()->isSubtypeOf(ListType::ofTensors()); });
 }
 
+// Returns if an operator is in-place. An operator is inplace if:
+// 1. its name ends in '_'
+// 2. The first argument is a Tensor and it is being written to
+// 3. The first argument is being returned
+// 4. No other arguments are aliased
+// Here is an example of an in-place operator:
+// add_(Tensor(a!) self, Tensor other, *, Scalar alpha=1) -> Tensor(a!)
+static bool isInplaceOp(const c10::FunctionSchema& schema) {
```

Comment:
You don't have to fix this here, but this should be precomputed in FunctionSchema, as the inplace-ness is of general interest to others. In fact, I have a version of this function in conjugate view tensors: https://github.com/pytorch/pytorch/pull/44799/files#diff-e0b7fe2a6c006c34bd33d0808fe1c9f4659d0faf27f58d64758c96be46c9ac77R31 though I am not only testing for inplaceness, I am testing for any mutation at all.

One Q I will be trying to answer as I read the rest of this PR is whether or not you really needed this algorithm to only work on inplace operations, or if any mutating operation would have been OK.

Reply:
The algorithm assumes that the first argument of the function is the only one that is being mutated. It is possible to extend the algorithm to support functions that mutate a different argument (like out=) or mutate multiple arguments, but those require some more bookkeeping (and the plan for vmap is to not support out= overloads)

Reply:
I'm happy to toss something like this into FunctionSchema, in this PR or in a separate PR

Reply:
out= doesn't seem like a hard generalization. Besides for safety checks, the mutating argument(s) simply just may occur in other positions.

Of course, boxed fallback for out functions doesn't work yet (it's buggy), so you couldn't implement it anyhow yet.